sw/device/lib/base/memory.c

```diff
@@ -12,26 +14,109 @@
+ * addresses. This function effectively breaks the given buffers into three
+ * consecutive chunks: (1) unaligned leading bytes, (2) aligned words, and (3)
+ * unaligned trailing bytes.
+ *
+ * @param[in] left The memory function's first buffer argument. Cannot be NULL.
+ * @param[in] right The memory function's second buffer argument. May be NULL.
+ * @param[in] len The length in bytes of both `left` and `right.`
+ * @param[out] out_leading_bytes_end The end of the region of leading bytes.
+ * @param[out] out_aligned_words_end The end of the region of aligned words.
+ */
+static void compute_alignment(uintptr_t left, uintptr_t right, size_t len,
+                              size_t *out_leading_bytes_end,
+                              size_t *out_aligned_words_end) {
+  const size_t num_leading_bytes =
+      compute_num_leading_bytes((uintptr_t)left, (uintptr_t)right, len);
+  *out_leading_bytes_end = num_leading_bytes;
+
+  const size_t num_words = (len - num_leading_bytes) / sizeof(uint32_t);
+  *out_aligned_words_end = num_leading_bytes + num_words * sizeof(uint32_t);
+
+#ifndef OT_PLATFORM_RV32
+  if (num_leading_bytes < len) {
+    assert(misalignment32_of(left + num_leading_bytes) == 0);
+    if (right != (uintptr_t)NULL) {
+      assert(misalignment32_of(right + num_leading_bytes) == 0);
+    }
+  }
+#endif
+}
+
+static uint32_t repeat_byte_to_u32(uint8_t byte) {
+  uint32_t extended = byte;
+  return (extended & 0xff) << 24 | (extended & 0xff) << 16 |
+         (extended & 0xff) << 8 | (extended & 0xff);
+}
+
 // The symbols below are all provided by the host libc. To avoid linker clashes,
 // they are all marked as `OT_WEAK`.
 
 OT_WEAK
 void *OT_PREFIX_IF_NOT_RV32(memcpy)(void *restrict dest,
                                     const void *restrict src, size_t len) {
-  uint8_t *dest8 = (uint8_t *)dest;
-  uint8_t *src8 = (uint8_t *)src;
-  for (size_t i = 0; i < len; ++i) {
-    dest8[i] = src8[i];
+  if (dest == NULL || src == NULL) {
+    return dest;
+  }
+  unsigned char *dest8 = (unsigned char *)dest;
+  const unsigned char *src8 = (const unsigned char *)src;
+  size_t leading_bytes_end, aligned_words_end;
+  compute_alignment((uintptr_t)dest, (uintptr_t)src, len, &leading_bytes_end,
+                    &aligned_words_end);
+  size_t index = 0;
+  for (; index < leading_bytes_end; ++index) {
+    dest8[index] = src8[index];
+  }
+  for (; index < aligned_words_end; index += sizeof(uint32_t)) {
+    uint32_t word = read_32(&src8[index]);
+    write_32(word, &dest8[index]);
+  }
+  for (; index < len; ++index) {
+    dest8[index] = src8[index];
   }
   return dest;
 }
 
 OT_WEAK
 void *OT_PREFIX_IF_NOT_RV32(memset)(void *dest, int value, size_t len) {
-  uint8_t *dest8 = (uint8_t *)dest;
-  uint8_t value8 = (uint8_t)value;
-  for (size_t i = 0; i < len; ++i) {
-    dest8[i] = value8;
+  unsigned char *dest8 = (unsigned char *)dest;
+  const uint8_t value_byte = (uint8_t)value;
```

Comment:
nit: just for consistency:
```suggestion
  const uint8_t value8 = (uint8_t)value;
```